src/System.IdentityModel.Tokens/AsymmetricSignatureProvider.cs

```diff
@@ -83,41 +83,58 @@ public AsymmetricSignatureProvider(AsymmetricSecurityKey key, string algorithm,
                 throw new ArgumentOutOfRangeException("key.KeySize", key.KeySize, string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10630, key.GetType(), SignatureProviderFactory.MinimumAsymmetricKeySizeInBitsForVerifying));
             }
 
+#if DNXCORE50
+            hash = GetHashAlgorithmName(algorithm);
+#else
             hash = GetHashAlgorithm(algorithm);
+#endif
             RsaSecurityKey rsaKey = key as RsaSecurityKey;
+
             if (rsaKey != null)
             {
+#if DNXCORE50
+                rsaCng = new RSACng();
+                (rsaCng as RSA).ImportParameters(rsaKey.Parameters);
+#else
                 rsaCryptoServiceProvider = new RSACryptoServiceProvider();
                 (rsaCryptoServiceProvider as RSA).ImportParameters(rsaKey.Parameters);
+#endif
                 return;    
             }
 
             X509SecurityKey x509Key = key as X509SecurityKey;
             if (x509Key != null)
             {
-                RSACryptoServiceProvider rsa = null;
+
+#if DNXCORE50
                 if (willCreateSignatures)
                 {
-					rsa = x509Key.PrivateKey as RSACryptoServiceProvider;
+                    rsaCng = RSACertificateExtensions.GetRSAPrivateKey(x509Key.Certificate) as RSACng;
```

Comment:
@PinpointTownes you're correct; there are cases where CNG can report back that it doesn't know how to handle the key (hardware keys with CAPI-only providers, for example), in which case it falls back to CAPI (RSACryptoServiceProvider).

Or, you know, you're running on non-Windows :).